experiment_prototype/scan_classes/sequences.py

```diff
@@ -337,107 +260,97 @@ def __init__(self, seqn_keys, sequence_slice_dict, sequence_interface, transmit_
 
+                # we have [p,e] and [a,s], but we want [p,a,(e*s)]. Adding null axis to encoding
+                # will produce this result.
+                phase_encoding = np.exp(1j * phase_encoding[:,np.newaxis,:])
+                samples = phase_encoding * basic_samples
 
-            if repeat:
-                pulse_antennas = []
             else:
-                # Initialize a list of lists for samples on all channels.
-                # TODO: modify this function if we put a weighting on powers instead of just a
-                # simple power_divider integer
-                pulse_samples, pulse_antennas = (
-                    make_pulse_samples(one_pulse_list, self.power_divider, self.slice_dict,
-                                       slice_to_beamdir_dict, txrate, txctrfreq,
-                                       main_antenna_count, main_antenna_spacing, pulse_ramp_time,
-                                       max_usrp_dac_amplitude, tr_window_time))
-                if pulse_index == 0:
-                    # calculate the first rx sample and set the value.
-                    self.first_rx_sample_time = calculate_first_rx_sample_time(
-                        pulse_samples[0].shape[0], txrate)
-                # Can plot for testing here
-                # plot_samples('channel0.png', pulse_samples[0])
-                # plot_fft('fftplot.png', pulse_samples[0], prog.txrate)
-
-            # This is all the data required for a pulse.
-            pulse_transmit_data['startofburst'] = startofburst
-            pulse_transmit_data['endofburst'] = endofburst
-            pulse_transmit_data['pulse_antennas'] = pulse_antennas
-            pulse_transmit_data['samples_array'] = pulse_samples
-            pulse_transmit_data['timing'] = timing
-            pulse_transmit_data['isarepeat'] = repeat
-
-            # Add pulse dictionary pulse_transmit_data at last place in sequence list
-            sequence_list.append(pulse_transmit_data)
-
-        self.find_blanks()
-
-        return sequence_list
+                samples = np.repeat(basic_samples[np.newaxis,:,:], num_pulses, axis=0)
+
+            # sum the samples into their position in the sequence buffer.
+            tr_window_num_samps = round(tr_window_time * txrate)
+            for i, pulse in enumerate(self.combined_pulses_metadata):
+                for component_info in pulse['component_info']:
+                    if component_info['slice_id'] == slice_id:
+                        pulse_timing_us = component_info['start_time_us']
+                        pulse_sample_start = round(txrate * (pulse_timing_us * 1e-6))
+                        pulse_samples_len = samples.shape[-1]
+
+                        start = tr_window_num_samps + pulse_sample_start
+                        end = start + pulse_samples_len
+                        pulse_piece = sequence[...,start:end]
+
+                        np.add(pulse_piece, samples[i], out=pulse_piece)
```

Comment:
this is nice. Adding all components in the correct place and the overlap doesn't need to be a special case. 

![](https://media.giphy.com/media/Kg9LGj9fgm3ImOg2Sp/giphy.gif)